Block Maxima

PythagoreanCult opened the pull request:
Includes functions to estimate the fractal dimension from data through the  Block Maxima method of extreme value theory

PythagoreanCult committed:
Add files via upload

PythagoreanCult committed:
Delete FractalDimensions.jl

PythagoreanCult committed:
add extremesdim

PythagoreanCult committed:
changed misspelled persistence

PythagoreanCult committed:
remove annoying folder
 Please enter the commit message for your changes. Lines starting

PythagoreanCult committed:
Merge branch 'Ignacio'

Datseris committed:
turn the initial string into comment

Datseris committed:
change the return type to Tuple

This is standard practice; functions input and output tuples. There is no reason to make the returned Tuple into a Vector, and if anything it is bad practice; it makes a type unstable vector and it serves no purpose. Anything you could do with this two element vector you could do better and faster with a Tuple.

Datseris committed:
Delete all `;`

No instance of `;` in the source code had reason to exist. Julia is capable of automatically deducing when a statement ends. Only reason for `;` is to have to commands on one line like:
`x = 3; y = x + 2`.

Datseris committed:
organize import and export statements

Datseris committed:
Correct structure in docstrings

Datseris committed:
replace quanti with q

Also, make the error messages much shorter

Datseris committed:
remove space from all `:: ` 

Apparently it is possible to have space between `::` but it is very un-idiomatic; noone does it in real code, which makes putting spaces hard to read.

Datseris committed:
replace `AbstractVector` with `AbstractStateSpaceSet`

This is arguably the most important change I do in this PR. Read the docstring of `StateSpaceSet` to understand why.

Datseris committed:
Fix the `extremal_index_sueveges` function...

There were a lot of problems:

- the function name was different from where it was called
- the function input arguments were named differently from where the function was called. This was confusing.
- The order of arguments and the documentation string where wrong. This was ULTRA confusing. The threshold and quantile arguments were exchanging places in the documentation.
- The quantile and threshold were named the same way. Not possible to deduce which was what besides going through the algorithm.

Datseris committed:
rename variables for clarity

Datseris committed:
don't call extremal index; function is wrong

Datseris committed:
use a normal way to get length of set

Datseris committed:
add convenience function for dim

Datseris committed:
first pass at performance: complete refactor of functions

The point is: local estimation should be its own function because it is a dedicated piece of algorithm. Then, we can optmize stuff by looping over each "column" of the distance matrix. If it fits in memory, we also gain massive BLAS-enabled performance. If not, we can still optimize by pre-allocating vectors and not re-allocated memory in each pass.

Datseris committed:
add convenience function for local values

Datseris committed:
add progress bar

and also stop using `AbstractStateSpaceSet` in function call signatures

Datseris committed:
add test file

Datseris committed:
correctly use greater-equals instead of only greater

Datseris committed:
add second estimator for pareto sigma

Datseris committed:
parallelize via threads

Datseris committed:
Merge branch 'main' into Ignacio

Datseris committed:
Merge branch 'main' into Ignacio

Datseris committed:
add preliminary tests for extreme value dim

Datseris committed:
fix test for extremal index

Datseris committed:
wip] add type for deciding estimator (?)

Datseris committed:
we don't need a type in the end.

Datseris committed:
Merge branch 'main' into Ignacio

Datseris committed:
rename q to p as it quantifies a probability, not the quantile

Datseris committed:
wip evt

Datseris committed:
finish docs EVT

Datseris committed:
add passing tests to EVT

Datseris committed:
add to docs

Datseris committed:
changelog

PythagoreanCult committed:
Merge branch 'Ignacio' of https://github.com/PythagoreanCult/FractalDimensions.jl into Ignacio

PythagoreanCult committed:
added the functions that estimate parameters for a GEV fit.

PythagoreanCult committed:
Added method to obtain the local dimensions and the persistances through
the Block Maxima approach

PythagoreanCult committed:
Changes requested, removed spaces from docstring and modified the
function to get the second moment to use less memory

PythagoreanCult committed:
Removed spaces from docstring

PythagoreanCult committed:
Removed space between docstring and function

PythagoreanCult committed:
The docstring is expandend and fixed according to the comments. Also the
dimension name is changed from D1 to Delta

PythagoreanCult committed:
An empty line is removed between the docstring and the function

PythagoreanCult committed:
Comments resolved: remove unhelpfull message, remove indexes from the
argument of the lenght function and remove the vector wrapping in the
return statement

PythagoreanCult committed:
Changed the variable name quanti--> p like in the rest of the code

PythagoreanCult committed:
Changed the floor function to give the correct type instead of calling
Int64 to turn argument into the desired type

PythagoreanCult committed:
Changed the floor function to return the correct type

PythagoreanCult committed:
Added absolute value function to prevent the domain errors in the square
root

PythagoreanCult committed:
Added progress bar

PythagoreanCult committed:
Added the input parameter blocksize to allow custom block size and
number of blocks. Added error handling for the case in which there is
duplicate points

PythagoreanCult committed:
Changed the docstring to include recent changes

PythagoreanCult committed:
Changed the error handling, created the struct BlockMaximaMM necessary for the
multiple dispatch and changed the function name

PythagoreanCult committed:
Added the changes to include structs BlockMaxima, and Exceedances for
the multiple dispatch.
Erased the matrix allocation functions and calls.

Datseris committed:
Merge branch 'main' into Block

Datseris committed:
Merge branch 'main' into Block

Datseris commented on src/extremes_based/gev.jl:
```diff
@@ -0,0 +1,40 @@
+    σ = √(abs(moment2-moment1^2)/(π^2/6))
+    return σ
+end
+export estimate_gev_scale
```
This function has no reason to exist. If you only need σ use the above function!

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -7,6 +7,19 @@ using Distances: euclidean
+    p::Real
+ end
+
+ struct ExceedancesMM
```
Delete this type in favor of the one below. Put the one below here.

Datseris reviewed (changes requested):
@PythagoreanCult it is difficult to review this PR as it is now, because it is not finished and it doesn't have tests. Please put the package in local development mode, using `dev` command, and then try the package and show whether it does what you expect. (It won't work, so you need to modify the code until it works) .You also need to add tests to the `test` folder that show how the new functionality is to be used, and whether it returns what you expect.

ps: Don't forget to first **pull** the changes I've pushed to the branch. Go to GitKraken and then click the "pull" button.

PythagoreanCult committed:
In the test file: Modified the tests to include the types of the new code. Include the

estimators as instances of structs and erased the test for wrong
estimator since is no longer relevant
In the extreme dims file: Update the functions so they no longer trigger errors on execution, erased the calls that still depended on estimator as a symbol instead of a struct.

PythagoreanCult committed:
Added docstrings for the structs and fixed a couple of errors regarding
the use of structs in the code, and the call to the function if the
argument is not a struct but a Real

PythagoreanCult committed:
Erased some comments to myself from the code

PythagoreanCult committed:
Updated the docstrings of several functions to reflect the changes to the struct based definition of the methods

PythagoreanCult committed:
Added back the test of the method :pwm as it was before

PythagoreanCult committed:
Fixed running errors

Fixed errors on code: typo `x`->`X` in both functions of the gev file and in the file extremesdim changed the definition of logdist in the first function because it rewrited the vector in a way that lost elements iteratively in combination whith erasing the Inf value

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -34,28 +71,21 @@ function extremevaltheory_dims(X, p; kw...)
-given set for quantile probability `p`, according to the estimation done via extreme value
-theory [Lucarini2016](@cite).
+given set. The type `p` tells the function which approach to use when computing the
+dimension, see `BlockMaxima` and `Exceedances`. The exceedances approach follows the
```
Instead of type `Name` , wrap this in an `@docs` reference block. Wrap the quoted name `Name` in [ ] and in `(@ref)`.

Datseris replied:
After this sentence you should also add the sentence "providing `p::Real` defaults to using the exceedances method".

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -34,28 +71,21 @@ function extremevaltheory_dims(X, p; kw...)
 
 """
-    extremevaltheory_dims_persistences(x::AbstractStateSpaceSet, p::Real; kwargs...)
+    extremevaltheory_dims_persistences(x::AbstractStateSpaceSet, p::type; kwargs...)
```
`type` doesn't exist. It is not a type that actually exists. It is better to write just `p` instead of `p::type`.

Datseris commented on src/extremes_based/extremesdim.jl:
```diff
@@ -34,28 +71,21 @@ function extremevaltheory_dims(X, p; kw...)
 Return the local dimensions `Δloc` and the persistences `θloc` for each point in the
-given set for quantile probability `p`, according to the estimation done via extreme value
-theory [Lucarini2016](@cite).
+given set. The type `p` tells the function which approach to use when computing the
```
```suggestion
given set. The type of `p` tells the function which approach to use when computing the
```

this is a crucial distinction. It is not `p` that tells, it is the _type of_ p.

Datseris reviewed (commented):
I am sorry reviewing this took so long. I have some eye problems.

This looks very good. It is a huge improvement. I have minor comments only. 

A problem we have is this: https://github.com/JuliaDynamics/FractalDimensions.jl/actions/runs/6652400509/job/18076301004?pr=30#step:7:429

it says that there is a method re-definition. We have to be careful about this. The method `extremevaltheory_local_dim_persistence` that dispatches on `::Exceedances` is indeed defined twice at lines 154 and 231. This is a mistake.

When you fix these minor comments I will merge this in. Is there any reference you can provide for the block maxima approach? Any paper that has used it so far to obtain a fractal dimension?

PythagoreanCult committed:
Adjusted the criteria of accuracy of the "Validity" test because the
test sometimes fails spuriously.

PythagoreanCult committed:
Fixed the docstring of the function extremevaltheory_dim() to include
doc references and correct the missuse of the word type.

PythagoreanCult committed:
Fixed the docstring of the function extremevaltheory_dims_persistances() to include the doc references and correct the misuse of the word type

PythagoreanCult committed:
Corrected the misuse of type as a type.

PythagoreanCult committed:
Erased the duplicated function

PythagoreanCult committed:
Added reference for the block maxima extreme value method.

Datseris committed:
Merge branch 'main' into Block

Datseris committed:
add the types to the documentation

Datseris committed:
separate docs into more blocks

Datseris committed:
simplify docstrings of convenience functions

Datseris committed:
separate types into their files

Datseris committed:
create more logical structuring in the files

Datseris committed:
fully separate code: dispatch happens in estimator files

Datseris committed:
improve, simplify, and cohere, the codstrings

Datseris committed:
changelog

Datseris reviewed (approved):
Great, thank you for your work @PythagoreanCult . I took the liberty to finish this PR by doing a bunch of small changes that have big impact on code clarity. You can go through the commits I've pushed one by one on your GitKtraken by clicking on the commit in the commit-tree-view. My commits do not change any code behavior. THey are about rte-writting code so taht it is better organized and easier to read. I also improved the documentation, and reduced the overall total amouint of documentation, following the rule "each word in the documentation must be weighted in gold"., 

I will merge once tests pass.

I also incremented minor version as this is a new feature. v1.8 is the new version.

Datseris committed:
add more code authors

Datseris committed:
add tests for deprecation
